azerothcore/azerothcore-wotlk
# fix(DB/Core) Fix the CLS damage system and update creature_classlevelstats

### Fix the CLS damage calculation system and update creature_classlevelstats accordingly

## Changes Proposed:
-  Reenable the min/max damage CLS system in Creature::SelectLevel in Creature.cpp that has been disabled
-  Fix the underlying calculation issue in Creature::CalculateMinMaxDamage in StatSystem.cpp. AttackMultiplier was wrongfully multiplied with AttackSpeed. I used the reverse engineered CLS calculations TC/vmangos/cmangos use.
- Correct several wrong values in creaure_classlevelstats. Multiple damage_base and attackpower values were wrong before. I imported those values from TC's database.
- Correct several wrong DamageModifier values in creature_template. Using TC values for those aswell. The values we had before were mostly a flat 1 or completly off because they were based on the old system. With the new values the calculations lead to the desired result.
- Drop "mindamage", "maxdamage", "minrangedamage",  "maxrangedamage", "attackpower" and "rangedattackpower" from creature_template as those values aren't needed anymore.
- Fix all code that references those columns.
- Implement BaseVariance and RangeVariance (mostly for custom mobs).



## Issues Addressed:
- Closes #4570 

- While no other issues are open at the moment, using the CLS calculation instead of using fixed min/max values could potentially fix a lot of NPCs which had wrong sniffed data before. This bug existed since Commit 1 of AzerothCore :)

## Tests Performed:
- Tested "reload creature_template".
- Used the Spell "Beast Lore (ID=1462) on around 20 different mobs to check their min/max damage values. Those values now correctly change if a mob is for example Level 35 vs Level 36. All values I tested so far have been ~99% to the pre-calculated min/max values so it seems like the new damage_base values are correct or atleast much better than the values we had before.
- Played for a few hours on my offline realm and everything I tried looks great.



## How to Test the Changes:
 - Download the PR, compile the code and apply the DB update.
 - Make a Hunter or simply learn spell 1462. 
 - Fly around the world and check the mobs min/max attack values.


## Known Issues and TODO List:
- I have no way to confirm the values I took from TC are 100% accurate. BUT they are way better than the precalculated values we used before and every mob I tested has been correct.


## Target Branch(es):

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR
